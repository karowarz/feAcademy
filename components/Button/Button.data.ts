import { ButtonProps } from './Button';

const common: ButtonProps = {
	label: 'This is Button',
};

const ButtonData: { [key: string]: ButtonProps } = {
	Base: common,
	Primary: { ...common, primary: true },
	Large: { ...common, size: 'large' },
	Small: { ...common, size: 'small' },
	Disabled: { ...common, disabled: true },
};

export default ButtonData;