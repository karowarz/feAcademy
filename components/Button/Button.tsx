/* eslint-disable no-mixed-spaces-and-tabs */
import React, { ButtonHTMLAttributes } from 'react';
import styles from './Buttton.module.scss';
import classNamesBind from 'classnames/bind';

const cx = classNamesBind.bind(styles);

export interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
	// Optional className
	className?: string;
	/**
	 * Is this the principal call to action on the page?
	 */
	primary?: boolean;
	/**
	 * What background color to use
	 */
	backgroundColor?: string;
	/**
	 * How large should the button be?
	 */
	size?: 'small' | 'medium' | 'large';
	/**
	 * Button contents
	 */
	label: string;
	/**
	 * Optional click handler
	 */
	onClick?: () => void;

	disabled?: boolean;
}

/**
 * Primary UI component for user interaction
 */
export const Button = ({
	                       className,
	                       primary = false,
	                       size = 'medium',
	                       disabled,
	                       backgroundColor,
	                       label,
	                       ...props
                       }: ButtonProps) => {
	const classNames = cx({
		'storybook-button': true,
		[`storybook-button--${size}`]: true,
		'storybook-button--primary': primary,
		'storybook-button--secondary': !primary,
		'storybook-button--disabled': disabled,
		[`${className}`]: !!className
	});

	return (
		<button
			type="button"
			className={classNames}
			style={{ backgroundColor }}
			{...props}
		>
			{label}
		</button>
	);
};
