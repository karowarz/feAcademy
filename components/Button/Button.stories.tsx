import React from 'react';
import { ComponentMeta, ComponentStory } from '@storybook/react';
import { Button } from './Button';
import ButtonData from './Button.data';

export default {
	title: 'Components/Button',
	component: Button,
	argTypes: {
		backgroundColor: { control: 'color' },
	},
} as ComponentMeta<typeof Button>;

const Template: ComponentStory<typeof Button> = (args) => <Button {...args} />;

export const Base = Template.bind({});
Base.args = ButtonData.Base;

export const Primary = Template.bind({});
Primary.args = ButtonData.Primary;

export const Large = Template.bind({});
Large.args = ButtonData.Large;

export const Small = Template.bind({});
Small.args = ButtonData.Small;

export const Disabled = Template.bind({});
Disabled.args = ButtonData.Disabled;
