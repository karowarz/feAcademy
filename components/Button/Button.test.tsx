/**
 * @jest-environment jsdom
 */

import { render } from '@testing-library/react';
import { Story } from '@storybook/react';
import { axe } from 'jest-axe';
import { ButtonProps } from './Button';
import * as stories from './Button.stories';

Object.entries(stories)
	.filter(([storyName]) => storyName.toLowerCase() !== 'default')
	.forEach(([storyName, componentStory]) => {
		const Component = componentStory as (Story<ButtonProps> & { args: ButtonProps });

		test(`renders ${storyName} without crashing`, async () => {
			const { container } = render(<Component {...Component.args}/>);

			//Snpashot testing
			expect(container).toMatchSnapshot();

			//axe a11y testing
			expect(await axe(container)).toHaveNoViolations();
		});
	});
