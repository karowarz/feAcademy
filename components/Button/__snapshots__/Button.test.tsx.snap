// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Base without crashing 1`] = `
<div>
  <button
    class="storybook-button storybook-button--medium storybook-button--secondary"
    type="button"
  >
    This is Button
  </button>
</div>
`;

exports[`renders Disabled without crashing 1`] = `
<div>
  <button
    class="storybook-button storybook-button--medium storybook-button--secondary storybook-button--disabled"
    type="button"
  >
    This is Button
  </button>
</div>
`;

exports[`renders Large without crashing 1`] = `
<div>
  <button
    class="storybook-button storybook-button--large storybook-button--secondary"
    type="button"
  >
    This is Button
  </button>
</div>
`;

exports[`renders Primary without crashing 1`] = `
<div>
  <button
    class="storybook-button storybook-button--medium storybook-button--primary"
    type="button"
  >
    This is Button
  </button>
</div>
`;

exports[`renders Small without crashing 1`] = `
<div>
  <button
    class="storybook-button storybook-button--small storybook-button--secondary"
    type="button"
  >
    This is Button
  </button>
</div>
`;
